### ISSUE TITLE
Cannot open macOS VM - limit exceeded

### ISSUE THREAD

**Issue by markbattistella** [utmapp/UTM#5563]
**BEFORE SUBMITTING YOUR ISSUE, PLEASE LOOK AT THE PINNED ISSUES AND USE THE SEARCH FUNCTION TO MAKE SURE IT IS NOT ALREADY REPORTED. ALWAYS COMMENT ON AN EXISTING ISSUE INSTEAD OF MAKING A NEW ONE.**

**Describe the issue**  

I have searched the previous issues and found nothing about the VM limits for running macOS on UTM. As I understand there is a hard limit imposed by Apple to not allow more than 2 macOS VMs to run.

The issue I face is that when I shutdown my VM via the VM system, and UTM lists the VM as offline, I can still not re-launch it.

I receive the error:

```
Error
The number of virtual machines exceeds the limit. The maximum supported number of active virtual machines has been reached.
```

I have no other VMs running, I have quit the UTM app, and even checked Activity Monitor. The only way to reset the "count" is a system reboot.

I haven't been able to find anything on fixing this, or if there is an issue at the moment with the way shutting down doesnt minus a count number.

**Configuration**  
* UTM Version: 4.3.5(87)
* macOS Version: 13.4.1 (c)
* Mac Chip (Intel, M1, ...): MacBook Pro M2 Max

**Comment by ideologysec** [comment]
Is Virtual Machine Service still running? This is the actual process that runs the virtual machine. If you force quit that, does the issue resolve?

<img width="338" alt=" 2023-08-15 at 23 37 35" src="https://github.com/utmapp/UTM/assets/790061/d9a66ffb-8e2e-4584-8fd8-ef52391f114e">

**Comment by markbattistella** [comment]
@ideologysec so I waited for the issue to occur again.

But, after using the VM all day yesterday today it now reports the error message.

### Chronology

**2023-08-18:**

- Used the macOS VM all day, internally rebooted a few times. No issues whatsoever

**Notes:** 

- physical mac machine did not go to sleep 
- physical mac machine did not have lid closed
- physical mac machine did have screen go off / black throughout day

**2023-08-18:**

- End of day, shutdown the macOS VM internally
- Quit the UTM app
- I put physical mac to sleep and closed the lid

**2023-08-19:**

- Launch UTM, and tried to boot the macOS VM
- Error message appears
- Check activity monitor: no `Virtual Machine Service` present, UTM app is present

**Notes:** 

- attempted to sleep physical mac to replicate issue: did not seem to cause issue
- attempted to close lid on physical mac to replicate issue: did not seem to cause issue
- attempted to sleep and close lid on physical mac to replicate issue: did not seem to cause issue

---

At this point I can't replicate it, it just occurs at times without any UTM or Virtualisation services running.

I don't have any knowledge on collecting logs or ways to see what is occurring when I get the error message.

**Comment by runwaygoats** [comment]
https://kb.parallels.com/128867

_running more than two such VMs simultaneously is impossible._

Needs to be verified if this is an Apple Virtualization Framework limit.

**Comment by ideologysec** [comment]
Yes. Apple has hardcoded into the released OS' a check for how many macOS VMs are running at one time.

Here's a technical deep dive with more info:
https://khronokernel.github.io/macos/2023/08/08/AS-VM.html

**Comment by howieisaacks** [comment]
For me, this is happening when NONE of my VMs are already running. It's now affecting both UTM and Parallels.

**Comment by pthexton** [comment]
This is the same issue as I raised with #4257  - I would suggest there may be other issues reported that are also basically the same problem and maybe a bit of house cleaning is needed to merge them / close duplicates.

Anyhow, as far as I've been able to tell the Mac firmware tracks the active number of VMs somehow but somewhere along the line this tracking gets it wrong and it thinks there are still macOS Guest VMs running even though there aren't. 

When I do run in to this issue (very rare now as I use the CLI `tart` utility instead of using UTM where I don't seem to run in to this problem), the easiest solution I've found is to log off the mac and back in again (a full reboot isn't required)

**Comment by danik85sh** [comment]
Same issue here with M1 MacBookPro running UTM 4.4.5(94) on MacOS Sonoma 14.4.
<img width="2315" alt="Screenshot 2024-03-13 at 10 49 57" src="https://github.com/utmapp/UTM/assets/38634449/f1799106-8779-49ad-b8c3-2b60188d3f7a">

**Comment by sunnyyoung** [comment]
try run `nvram 40A0DDD2-77F8-4392-B4A3-1E7304206516:boot-args='hv_apple_isa_vm_quota=0xFF'` and reboot 😉

**Comment by sunnyyoung** [comment]
@ideologysec 

Sure!

My environments:

- Mac Mini M1 & Mac Mini M4
- macOS Sequoia 15.4 (24E248)
- [Tart](https://tart.run) VM

just run `sudo nvram 40A0DDD2-77F8-4392-B4A3-1E7304206516:boot-args='hv_apple_isa_vm_quota=0xFF'` in terminal, no recovery mode needed, and reboot.

the `40A0DDD2-77F8-4392-B4A3-1E7304206516` is nvram's partition identifier constant value, you should not change it 🤔, and the `hv_apple_isa_vm_quota=0xFF` is the key flag for breaking through the limitation.

I tested exactly as https://khronokernel.com/macos/2023/08/08/AS-VM.html, it works, then I perform `bputil --full-security` to restore and it still works 🤔, so I tried to streamline the process even further, I fully reset my devices with ipsw file and just tried `nvram 40A0DDD2-77F8-4392-B4A3-1E7304206516:boot-args='hv_apple_isa_vm_quota=0xFF'` and it still works perfectly 🤩, but I don't know if this is intentional or a bug in the latest macOS system.

**Comment by bwspot** [comment]
Running into same issue.  Vm UTM Mac works for some time and then out of no where after shutdown it shows the limit, although no vms are running.  Any fix for this without logging off?

**Comment by KStartups** [comment]
> Last I knew the `hv_apple_isa_vm_quota` argument is only valid on a Development or patched kernel. Same article I linked above:
> 
> https://khronokernel.com/macos/2023/08/08/AS-VM.html
> 
> For the sake of it I tried your command, the UUID doesn't seem to matter. Presumably that's your boot-volume UUID? It did not allow bypassing the 2 VM limit in Sequoia 15.4.
> 
> Can you share more details about this? Your OS? where you found the command and what it should do? [@sunnyyoung](https://github.com/sunnyyoung)

Will this work on Sequoia? @ideologysec @sunnyyoung Has anyone tried this? I need to urgently run more than 5 VMs on my M4 Sequoia but this 2 vms limit is killing me

**Comment by ideologysec** [comment]
> Will this work on Sequoia? [@ideologysec](https://github.com/ideologysec) [@sunnyyoung](https://github.com/sunnyyoung) Has anyone tried this? I need to urgently run more than 5 VMs on my M4 Sequoia but this 2 vms limit is killing me

If you're willing to go through the trouble of acquiring the KDK and building your own kernel and downgrading boot security, I see no reason why it wouldn't theoretically work - but I have not personally tested it with any recent OS'

Let's test it.

**Comment by sunnyyoung** [comment]
@KStartups It works on Sequoia, that's what I'm using it with now, but the last KDK you can get is 15.6(24G84), so this is the last usable system version.

**Comment by ideologysec** [comment]
@sunnyyoung try this maybe?

https://github.com/dortania/KdkSupportPkg/releases

**Comment by sunnyyoung** [comment]
@ideologysec It appears the data source is https://developer.apple.com/download/all/ . I've confirmed that Apple hasn't released a formal version of the KDK since version 15.6, but Beta/RC versions do.

**Comment by daemontroll** [comment]
Is there a way to decrement or clear the running vm counter?
